### PR TITLE
Refactor optimal filtering

### LINKED ISSUE #312: refactor filter code
The filter code is written in poor style in a variety of ways
1. it relies on a lot of internal mutation, making it hard to follow
2. it keeps dictionaries of filter values and variances for different flavors of filters that vary only by what they are orthogonal
3. because of 2 the code is hard to follow, and features a lot of repetition

Here is an example (incomplete) refactoring that avoid the use of mutation, limits code repetition, and generalizes over the differences between filters (noconst, baseline, etc). Moving the code in this direction is desireable. Now all the core math lives in just a few lines in `compute_filter` and `compute_fourier_filter`.

```python
from dataclasses import dataclass, field
import numpy as np
from scipy.linalg import toeplitz


@dataclass(frozen=True)
class Filter:
    """A computed filter with metadata about its creation."""
    values: np.ndarray
    variance: float
    factory: "FilterFactory"
    modification_fn: callable

    def __repr__(self):
        return (f"Filter(values=array(shape={self.values.shape}), variance={self.variance}, "
                f"factory=FilterFactory(avg_signal=array(shape={self.factory.avg_signal.shape}), "
                f"modification_fn={self.modification_fn.__name__}))")

    @property
    def vdv(self):
        """Compute the signal-to-noise ratio (SNR) of the filter."""
        # Signal: Use the norm of the filter values as the "signal"
        signal = np.linalg.norm(self.values)

        # Noise: The variance of the filter is used as the "noise"
        noise = self.variance

        # Signal-to-noise ratio (SNR)
        return signal / np.sqrt(noise) if noise != 0 else np.inf


@dataclass(frozen=True)
class FilterFactory:
    """Factory to compute filters."""
    avg_signal: np.ndarray
    n_pretrigger: int
    sample_time_s: float
    noise_autocorr: np.ndarray = None  # Optional
    noise_psd: np.ndarray = None  # Required for Fourier-based filter
    whitener: callable = None

    def compute_filter(self, modification_fn):
        """Compute a filter using the provided modification function."""
        n = len(self.avg_signal)
        noise_corr = self.noise_autocorr[:n] if self.noise_autocorr is not None else np.ones(n)  # Default to ones
        TS = ToeplitzSolver(noise_corr, symmetric=True)
        Rinv_sig = TS(self.normalized_signal(self.avg_signal))
        Rinv_1 = TS(np.ones(n))

        # Use the modification function to create the filter values
        filter_values = modification_fn(Rinv_sig, Rinv_1)
        variance = np.dot(filter_values, Rinv_sig)

        return Filter(
            values=filter_values,
            variance=variance,
            factory=self,
            modification_fn=modification_fn,
        )

    def compute_fourier_filter(self):
        """Compute a filter using Fourier methods with noise_psd."""
        if self.noise_psd is None:
            raise ValueError("noise_psd must be provided for compute_fourier_filter.")

        n = len(self.avg_signal)
        normalized_signal = self.normalized_signal(self.avg_signal)

        # Compute the Fourier transform of the signal and the noise power spectral density
        signal_fft = np.fft.fft(normalized_signal)
        noise_psd_fft = np.fft.fft(self.noise_psd[:n])

        # Compute the filter in the frequency domain
        fourier_filter = np.conj(signal_fft) / (noise_psd_fft + 1e-6)  # Avoid division by zero

        # Transform back to the time domain
        filter_values = np.fft.ifft(fourier_filter).real

        # Compute the variance of the filter
        variance = np.dot(filter_values, normalized_signal)

        return Filter(
            values=filter_values,
            variance=variance,
            factory=self,
            modification_fn=self.compute_fourier_filter,
        )

    def shorten(self, front_samples: int, back_samples: int):
        """Return a new FilterFactory with truncated signal and compatible noise."""
        # Truncate the signal from the front and the back
        shortened_signal = self.avg_signal[front_samples:-back_samples]
        
        # Renormalize the shortened signal
        truncated_factory = FilterFactory(
            avg_signal=shortened_signal,
            n_pretrigger=self.n_pretrigger,
            sample_time_s=self.sample_time_s,
            noise_autocorr=self.truncate_noise(self.noise_autocorr, front_samples, back_samples),
            noise_psd=self.truncate_noise(self.noise_psd, front_samples, back_samples),
            whitener=self.whitener,
        )
        return truncated_factory

    def truncate_noise(self, noise, front_samples, back_samples):
        """Helper function to truncate noise_autocorr and noise_psd in a compatible way."""
        if noise is not None:
            return noise[front_samples:-back_samples]
        return None

    @staticmethod
    def normalized_signal(avg_signal, n_pretrigger):
        """Class method to compute the normalized signal."""
        pre_avg = avg_signal[:n_pretrigger - 1].mean()
        a = avg_signal.min()
        b = avg_signal.max()
        is_negative = pre_avg - a > b - pre_avg
        peak_signal = a - pre_avg if is_negative else b - pre_avg
        normalized_signal = (avg_signal - pre_avg) / peak_signal
        normalized_signal[:n_pretrigger] = 0.0
        return normalized_signal

    @staticmethod
    def baseline_modification(Rinv_sig, Rinv_1):
        """Modification function for baseline subtraction."""
        return Rinv_sig - (Rinv_sig.sum() / Rinv_1.sum()) * Rinv_1

    @staticmethod
    def noconst_modification(Rinv_sig, Rinv_1):
        """Modification function for no baseline subtraction."""
        return Rinv_1.sum() * Rinv_sig - Rinv_sig.sum() * Rinv_1


class ToeplitzSolver:
    """Helper to solve Toeplitz systems."""
    def __init__(self, autocorr, symmetric=True):
        self.autocorr = autocorr
        self.symmetric = symmetric
        self.matrix = toeplitz(autocorr)

    def __call__(self, vector):
        """Solve the system with the given vector."""
        return np.linalg.solve(self.matrix, vector)
```

with example usage

```python
# Example data
avg_signal = np.array([0.1, 0.2, 0.5, 1.0, 0.8, 0.3])
n_pretrigger = 2
sample_time_s = 0.01
noise_autocorr = np.array([1.0, 0.5, 0.2, 0.1])  # Optional
noise_psd = np.array([0.1, 0.2, 0.3, 0.4, 0.5, 0.6])  # Required for Fourier filter

# Create a factory
factory = FilterFactory(avg_signal, n_pretrigger, sample_time_s, noise_autocorr=noise_autocorr, noise_psd=noise_psd)

# Compute filters
baseline_filter = factory.compute_filter(FilterFactory.baseline_modification)
noconst_filter = factory.compute_filter(FilterFactory.noconst_modification)

# Print signal-to-noise ratio (vdv)
print(f"Baseline filter vdv: {baseline_filter.vdv}")
print(f"No constant filter vdv: {noconst_filter.vdv}")

```

### PR DESCRIPTION
**Pull request by joefowler**
Redesign optimal filtering code in line with the intentions of #312. Specifically:

* No object contains more than one type of filter.
* Remove all internal mutation. Instead, the optimal filter class becomes two, both of which are frozen dataclasses (i.e., their attributes cannot be changed after construction). 
* One class is a `FilterMaker` that knows the signal and noise model, and one is a `Filter`, which (of course) the `FilterMaker` can generate. The filter values, bandpass, model, etc of a `Filter` are fixed when it's created.
   * The `FilterMaker` can generate filters created in the traditional Fourier way, or in the time domain. 
   * The latter can be 5-lag or "Arrival-Time Safe" versions. 
   * In all cases, the filter can be created with a fixed number of samples at the start and/or end of a record given zero weight (a simple way to mimic the effect of having written shorter records).
   * Filters can also be created with a `f_3db` and/or `fmax` to low-pass filter the filter values, either with a 1-pole rolloff or with a strict band limit.
   * The "recompute" method is removed. If a filter needs (say) a new 3dB rolloff, then a new `Filter` object must be created.
* V/dV and variance estimates are still fixed, as was recently required in #307 / PR #311.
* Change HDF5 representation of filters (probably breaks old filters, or fails to load some aspects)
* Removes broken `ExperimentalFilter` code, though it would be straightforward to add Alpert's constrained filters in the future.
* Also uses modern type hints correctly and standardized doctstring style.

Fixes #312.

### CODE REVIEW

File: mass/mathstat/toeplitz.py [hunk 1]
```diff
@@ -154,14 +154,17 @@ def __precompute_symmetric(self):
 
     def __solve_symmetric(self, y):
```
Comment (ggggggggg) [inline mass/mathstat/toeplitz.py:155]: what argument type must y have since it has to have the `ndim` field? An `ndarray`? what number of dimensions does it need in the `y.ndim != 2` case?

File: tests/core/test_core.py [hunk 2]
```diff
@@ -444,14 +445,13 @@ def test_pulse_model_and_ljh2off(self, tmp_path_factory):
 
         should_be_identity = np.matmul(pulse_model.projectors, pulse_model.basis)
         wrongness = np.abs(should_be_identity - np.identity(n_basis))
-        # ideally we could set this lower, like 1e-9, but the linear algebra needs more work
+        print(f"Wrongness matrix (abs-max is {np.amax(wrongness)})")
         print(wrongness)
-        print(np.amax(wrongness))
-        assert np.amax(wrongness) < 0.16
+        assert np.amax(wrongness) < 1e-9
```
Comment (ggggggggg) [inline tests/core/test_core.py:450]: We have achieved the ideal!??

File: tests/mathstat/test_power_spectrum.py [hunk 3]
```diff
@@ -27,35 +26,3 @@ def test_values():
     expected = np.linspace(0, 0.5, 6)
     for a, b in zip(f, expected):
         assert a == pytest.approx(b)
-
-
-# test functions used in the oct23 true bq analysis
-# where the pulses did not live in ljh files, instead the pulses needed to be passed
-# explicitly as arguments to the psd and autocorr calculation
-def test_orthognal_to_exponential_filter():
```
Comment (ggggggggg) [inline tests/mathstat/test_power_spectrum.py:35]: I don't mind removing the Experimental filter, but I would like to keep a test that shows how to go from pulses and noise records to filters.

File: mass/core/optimal_filtering.py [hunk 4]
```diff
@@ -760,3 +169,529 @@ def W(self, N):
...
+        sig_ft[freq > fmax] = 0.0
+    if f_3db is not None:
+        sig_ft /= (1. + (1.0 * freq / f_3db)**2)
+
+    # n=filt_length is needed when filt_length is ODD
+    vector[:] = np.fft.irfft(sig_ft, n=filt_length)
+
+
+@dataclass(frozen=True)
+class Filter:
```
[the remark this answers is not in the document]
Comment (joefowler) [reply mass/core/optimal_filtering.py:216]: I think we could make this happen, yes. Can't decide whether to make that part of this PR or a future one.

Also, as we discussed, the Fourier filter really is type of 5-lag filter, but with a different way to compute it.

File: tests/core/test_core.py (hunk 2 above)
Comment (joefowler) [reply tests/core/test_core.py:450]: Apparently, yes. I don't know why this was broken before, but it sure isn't broken now.

File: mass/core/optimal_filtering.py [hunk 5]
```diff
@@ -1,669 +1,74 @@
...
-            axis2.plot(self.filt_baseline_pretrig, color='blue')
-        except AttributeError:
-            pass
-        try:
-            axis1.plot(self.filt_fourier, color='gold')
-        except AttributeError:
-            pass
-
-
 class ToeplitzWhitener:
```
[the remark this answers is not in the document]
Comment (joefowler) [reply mass/core/optimal_filtering.py:14]: Good point. It was actually easy, so when I push next set of updates (to this PR), this is done.

File: mass/mathstat/toeplitz.py (hunk 1 above)
Comment (joefowler) [reply mass/mathstat/toeplitz.py:155]: Good point. I have added type annotations and reworked this file, too.

File: mass/core/optimal_filtering.py [hunk 6]
```diff
@@ -760,3 +169,529 @@ def W(self, N):
...
+    n_pretrigger : int
+        The number of leading samples in the average signal that are considered
+        to be pre-trigger samples.  The avg_signal in this section is replaced by
+        its constant averaged value before creating filters.  Also, one filter
+        (filt_baseline_pretrig) is designed to infer the baseline using only
+        `n_pretrigger` samples at the start of a record.
+    noise_autocorr : Optional[npt.ArrayLike]
+        The autocorrelation function of the noise, where the lag spacing is
+        assumed to be the same as the sample period of `avg_signal`.  If None,
+        then several filters won't be computed.  (One of `noise_psd` or
```
[the remark this answers is not in the document]
Comment (joefowler) [reply mass/core/optimal_filtering.py:318]: Yes, fixed the doc-string in next commits, correcting the statement AND speaking to the correct units for noise info.

File: mass/core/optimal_filtering.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (joefowler) [reply mass/core/optimal_filtering.py:228]: Yeah, this PR is already big enough I decided to decline this idea. But it's next: see PR #314.

File: mass/core/channel.py [hunk 7]
```diff
@@ -747,65 +747,55 @@ def __load_filters_from_hdf5(self, overwrite=False):
...
-            raise Exception(f"filter_type={filter_type}, must be `ats` or `5lag`")
-        self.filter.fmax = fmax
-        self.filter.f_3db = f_3db
-        self._filter_type = filter_type
+                modelpeak = np.max(self.average_pulse)
+                maker = FilterMaker(self.average_pulse[:], self.nPresamples - self.pretrigger_ignore_samples,
+                                    self.noise_autocorr, self.noise_psd,
+                                    sample_time=self.timebase, peak=modelpeak)
+                if filter_type == "5lag":
+                    self.filter = maker.compute_5lag(fmax=fmax, f_3db=f_3db)
```
[the remark this answers is not in the document]
Comment (joefowler) [reply mass/core/channel.py:792]: The case that skips the FilterMaker is the one where enough stored HDF5 attributes are present that we can safely skip straight to the end goal: a Filter with all appropriate meta-data.

Is that a good enough answer?

[the remark this answers is not in the document]
Comment (ggggggggg) [reply mass/core/channel.py:792]: I'll accept it. I didn't realize we could save enough to calculate a filter but not enough to just say "this is the filter".

[the remark this answers is not in the document]
Comment (joefowler) [reply mass/core/channel.py:792]: Sometimes we do, sometimes we don't. Sometimes we directly store the metadata required to create a `Filter` object (like predicted variance and V/dV). Other times we store enough to compute it but not the required metadata itself.

It appears to depend both on filter type (5-lag vs ATS) and on which version of the HDF5 backup we are using. Ugly, but so it is. In future pipelines, we won't use HDF5-backed python objects, right?

File: mass/core/optimal_filtering.py [hunk 8]
```diff
@@ -760,3 +169,529 @@ def W(self, N):
...
+        An object that can make a variety of optimal filters, assuming a single signal and noise analysis.
+    """
+
+    signal_model: npt.ArrayLike
+    n_pretrigger: int
+    noise_autocorr: Optional[npt.ArrayLike] = None
+    noise_psd: Optional[npt.ArrayLike] = None
+    dt_model: Optional[npt.ArrayLike] = None
+    whitener: Optional[ToeplitzWhitener] = None
+    sample_time: float = 0.0
```
[the remark this answers is not in the document]
Comment (joefowler) [reply mass/core/optimal_filtering.py:346]: I took a long time to decide whether to argue with you about this, but I decided to do it your way. See 8e059ad6 (to be pushed soon).

File: tests/mathstat/test_power_spectrum.py (hunk 3 above)
Comment (joefowler) [reply tests/mathstat/test_power_spectrum.py:35]: Good idea. Done--see e1f27b7b

Comment (joefowler) [reply tests/mathstat/test_power_spectrum.py:35]: and f0008b97, with the test function renamed to have an accurate description.

File: mass/core/optimal_filtering.py [hunk 9]
```diff
@@ -302,23 +314,36 @@ class FilterMaker:
...
-        assumed to be the same as the sample period of `avg_signal`.  If None,
-        then several filters won't be computed.  (One of `noise_psd` or
-        `noise_autocorr` must be a valid array, or `whitener` must be given.)
+        assumed to be the same as the sample period of `avg_signal`.
     noise_psd : Optional[npt.ArrayLike]
-        The noise power spectral density.  If None, then filt_fourier won't be
-        computed.  If not None, then it must be of length (2N+1), where N is the
-        length of `avg_signal`, and its values are assumed to cover the non-negative
-        frequencies from 0, 1/Delta, 2/Delta,.... up to the Nyquist frequency.
+        The noise power spectral density.  If not None, then it must be of length (2N+1),
```
Comment (ggggggggg) [inline mass/core/optimal_filtering.py:319]: I like this documentation change. It's much easier to understand!

**Review by ggggggggg — APPROVED** [review]
